ntp-proto/src/nts/messages.rs

```diff
@@ -0,0 +1,1849 @@
+use std::borrow::Cow;
+
+use tokio::io::{AsyncRead, AsyncWrite};
+
+#[cfg(feature = "nts-pool")]
+use crate::nts::AlgorithmDescription;
+#[cfg(feature = "nts-pool")]
+use crate::packet::Cipher;
+
+use super::record::NtsRecord;
+use super::{AeadAlgorithm, ErrorCode, NextProtocol, NtsError, WarningCode};
+
+pub enum Request<'a> {
+    KeyExchange {
+        algorithms: Cow<'a, [AeadAlgorithm]>,
+        #[cfg_attr(feature = "__internal-fuzz", allow(private_interfaces))]
+        protocols: Cow<'a, [NextProtocol]>,
+        #[cfg(feature = "nts-pool")]
+        denied_servers: Cow<'a, [Cow<'a, str>]>,
+    },
+    #[cfg(feature = "nts-pool")]
+    FixedKey {
+        c2s_key: Box<dyn Cipher>,
+        s2c_key: Box<dyn Cipher>,
+        algorithm: AeadAlgorithm,
+        #[cfg_attr(feature = "__internal-fuzz", allow(private_interfaces))]
+        protocol: NextProtocol,
+    },
+    #[cfg(feature = "nts-pool")]
+    Support {
+        wants_protocols: bool,
+        wants_algorithms: bool,
+    },
+}
+
+impl Request<'_> {
+    pub async fn parse(mut reader: impl AsyncRead + Unpin) -> Result<Self, NtsError> {
```

Comment:
I don't really like how different this function is with or without the nts-pool cfg, but I don't really have an alternative either. Maybe you can put the section after the loop in its own function?

Reply:
The differences between with or without pool are quite unavoidable. As for the length of the function, I'll have a think about refactoring this further to reduce that, but I don't think splitting the loop and the part after is the right way to go given the amount of data flowing between the two. But let's do that after merging this.